src/graph.rs

```diff
@@ -86,6 +92,131 @@ impl<'tree> Graph<'tree> {
 
         DisplayGraph(self, ctx)
     }
+
+    pub fn display_json<'a>(&'a self, ctx: &'a Context) -> () {
+        // TODO: move this all into a new module
+        struct JSONGraph<'a, 'tree>(&'a Graph<'tree>, &'a Context);
+        struct JSONNode<'a>(usize, &'a GraphNode, &'a Context);
+        struct JSONEdge<'a>(&'a u32, &'a Edge, &'a Context);
+        struct JSONAttributes<'a>(&'a Attributes, &'a Context);
+        struct JSONIdentifier<'a>(&'a Identifier, &'a Context);
+        struct JSONValue<'a>(&'a Value, &'a Context);
```

Comment:
The `display_with` machinery works by creating a new ~~typecla~~—ahem, **_trait_**—which provides some context data to pass around for resolving interned strings. We require the same context for the same reason, but it was most convenient to follow the approach taken by `display_with`, above (but folded in this PR view): define a struct type locally and give it an implementation of `serde::Serialize`.

Because a `Serializer`'s methods call into its own API, however, we can't reasonably distribute the construction of these `serialize` implementations around all of the types in question behind some convenient (and useful in isolation) method like `display_with`. For example, calling `serialize_seq` returns a serializer with a `serialize_element` method which takes values with a `Serialize` trait—so it only allows for top-down folds, not bottom-up. We could still provide a `serialize_with` method for each of the public types which, constructing the corresponding private JSON struct and calling into it. But we wouldn't be able to recur via that method.

Thus we're somewhat constrained in the precise pattern we employ (and thus in how easy it will be to test this monstrosity), but at least we don't have to keep all of these types inside this one method 😄